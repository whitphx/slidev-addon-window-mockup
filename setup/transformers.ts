import type {
  MarkdownTransformContext,
  TransformersSetup,
} from "@slidev/types";

function WindowedCodeblock(ctx: MarkdownTransformContext) {
  ctx.s.replace(
    /^```(\S+)? *(\{[^\n]*\})? +window\n([\s\S]+?)\n```/gm,
    (full, lang = "", options = "", code = "") => {
      return (
        "<WindowMockup codeblock>\n" +
        "```" +
        lang +
        options +
        "\n" +
        code +
        "\n```" +
        "\n</WindowMockup>"
      );
    },
  );
}

// We avoid using `import { defineTransformersSetup } from "@slidev/types"`
// because `@slidev/types` can be not resolvable in the installed environment (maybe the monorepo case?).
// So we just import types and use the static type annotation here,
// instead of importing the actual implementation of the helper function.
const setup: TransformersSetup = () => {
  return {
    pre: [],
    preCodeblock: [WindowedCodeblock],
    postCodeblock: [],
    post: [],
  };
};

export default setup;
