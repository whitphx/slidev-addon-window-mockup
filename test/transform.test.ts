import { transformWindowedCodeblock } from "../setup/transformers";
import { test, expect } from "vitest";
import { createTransformContext } from "./_tutils";

test("language name + window", () => {
  const ctx = createTransformContext(`
\`\`\`js window
console.log('Hello, World!');
\`\`\`
`);

  transformWindowedCodeblock(ctx);
  ctx.s.commit();

  expect(ctx.s.toString()).toMatchSnapshot();
});

test("language name + ranges + window", () => {
  const ctx = createTransformContext(`
\`\`\`js {*|1|2} window
console.log('Hello, World!');
\`\`\`
`);

  transformWindowedCodeblock(ctx);
  ctx.s.commit();

  expect(ctx.s.toString()).toMatchSnapshot();
});

test("language name + ranges + options + window", () => {
  const ctx = createTransformContext(`
\`\`\`js {*|1|2}{at: 1} window
console.log('Hello, World!');
\`\`\`
`);

  transformWindowedCodeblock(ctx);
  ctx.s.commit();

  expect(ctx.s.toString()).toMatchSnapshot();
});

test("language name + ranges + options + window + windowOptions", () => {
  const ctx = createTransformContext(`
\`\`\`js {*|1|2}{'maxHeight': '100px'} window {'color': 'dark'}
console.log('Hello, World!');
\`\`\`
`);

  transformWindowedCodeblock(ctx);
  ctx.s.commit();

  expect(ctx.s.toString()).toMatchSnapshot();
});

test("language name + title + ranges + options + window + windowOptions", () => {
  const ctx = createTransformContext(`
\`\`\`js [foo.js] {*|1|2}{'maxHeight': '100px'} window {'color': 'dark'}
console.log('Hello, World!');
\`\`\`
`);

  transformWindowedCodeblock(ctx);
  ctx.s.commit();

  expect(ctx.s.toString()).toMatchSnapshot();
});
