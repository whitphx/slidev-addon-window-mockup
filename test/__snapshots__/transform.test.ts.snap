// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`language name + ranges + options + window + windowOptions 1`] = `
"
<WindowMockup codeblock v-bind="{'color': 'dark'}">
\`\`\`js {*|1|2}{'maxHeight': '100px'} 
console.log('Hello, World!');

\`\`\`
</WindowMockup>
"
`;

exports[`language name + ranges + options + window 1`] = `
"
<WindowMockup codeblock v-bind="{}">
\`\`\`js {*|1|2}{at: 1} 
console.log('Hello, World!');

\`\`\`
</WindowMockup>
"
`;

exports[`language name + ranges + window 1`] = `
"
<WindowMockup codeblock v-bind="{}">
\`\`\`js {*|1|2} 
console.log('Hello, World!');

\`\`\`
</WindowMockup>
"
`;

exports[`language name + title + ranges + options + window + windowOptions 1`] = `
"
<WindowMockup codeblock v-bind="{'title':'foo.js','color': 'dark'}">
\`\`\`js {*|1|2}{'maxHeight': '100px'} 
console.log('Hello, World!');

\`\`\`
</WindowMockup>
"
`;

exports[`language name + window 1`] = `
"
<WindowMockup codeblock v-bind="{}">
\`\`\`js
console.log('Hello, World!');

\`\`\`
</WindowMockup>
"
`;
